python/lsst/daf/butler/direct_butler/_direct_butler.py

```diff
@@ -1623,54 +1617,29 @@ def _ingest_file_datasets(
                     )
 
                 groupedDataIds[group_key][ref.dataId] = dataset
+            refs.extend(dataset.refs)
 
-            if existingRefs:
-                if len(dataset.refs) != len(existingRefs):
-                    # Keeping track of partially pre-existing datasets is hard
-                    # and should generally never happen. For now don't allow
-                    # it.
-                    raise ConflictingDefinitionError(
-                        f"For dataset {dataset.path} some dataIds already exist"
-                        " in registry but others do not. This is not supported."
-                    )
-
-                # Store expanded form in the original FileDataset.
-                dataset.refs = existingRefs
-            else:
-                groupedData[group_key].append(dataset)
-
-        # Now we can bulk-insert into Registry for each DatasetType.
-        for (datasetType, this_run), grouped_datasets in progress.iter_item_chunks(
-            groupedData.items(), desc="Bulk-inserting datasets by type"
-        ):
-            refs_to_import = []
-            for dataset in grouped_datasets:
-                refs_to_import.extend(dataset.refs)
-
-            n_refs = len(refs_to_import)
-            _LOG.verbose(
-                "Importing %d ref%s of dataset type %r into run %r",
-                n_refs,
-                "" if n_refs == 1 else "s",
-                datasetType.name,
-                this_run,
-            )
+        # Bulk insert.
+        grouped_refs, dimension_records, _ = self._prepare_for_import_refs(
+            self,
+            refs,
+            skip_missing=False,
+            register_dataset_types=True,
+            dry_run=dry_run,
+            transfer_dimensions=False,
+        )
 
-            # Import the refs and expand the DataCoordinates since we can't
-            # guarantee that they are expanded and Datastore will need
-            # the records.
-            imported_refs = self._registry._importDatasets(refs_to_import, expand=True)
+        self._import_dimension_records(dimension_records, dry_run)
+        imported_refs = self._import_grouped_refs(
+            grouped_refs, self, progress, dry_run=dry_run, expand_refs=True
```

Comment:
`progress` here feels a bit invasive, in a sense that it is not essential for what the method's primary purpose is. Could it be made optional and keyword-only?

Reply:
Ok. Will take another look. I was mostly trying to split the code up without changing it much. Maybe I could pass in the progress bar name instead.

Reply:
Looking at the code the progress is used in the file ingest and passed into the import prepare so passing in the progress bar name doesn't seem right. This is an internal API so I'm not sure what I gain by making it optional if I'm going to define it in the two places that call the method.